src/Bicep.Core/TypeSystem/DeferredTypeReference.cs

```diff
@@ -6,13 +6,13 @@ namespace Bicep.Core.TypeSystem
 {
     public class DeferredTypeReference : ITypeReference
     {
-        private readonly Func<TypeSymbol> typeGetterFunc;
+        private readonly Lazy<TypeSymbol> lazyType;
 
         public DeferredTypeReference(Func<TypeSymbol> typeGetterFunc)
         {
-            this.typeGetterFunc = typeGetterFunc;
+            lazyType = new(typeGetterFunc);
         }
```

Comment:
Copy+pasted from other comment:
> Also - we have caching in our declared & assigned type managers - are you sure that caching is also necessary here?

Reply:
I did see `typeGetterFunc` getting invoked multiple times (for example, if a type alias is reused for multiple properties of an object type or as the type for multiple parameters). An `ObjectType` may contain a deferred type reference for each property that uses a type alias as its type, and the same is true of `TupleType` instances. The type managers will cache the `TypeSymbol` that contains a deferred type reference, but the return value of `typeGetterFunc` is not currently cached.

Reply:
Makes sense!